pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java

```diff
@@ -655,12 +655,67 @@ protected void internalDeletePartitionedTopic(AsyncResponse asyncResponse, boole
         });
     }
 
-    protected void internalUnloadTopic(boolean authoritative) {
+    protected void internalUnloadTopic(AsyncResponse asyncResponse, boolean authoritative) {
         log.info("[{}] Unloading topic {}", clientAppId(), topicName);
         if (topicName.isGlobal()) {
             validateGlobalNamespaceOwnership(namespaceName);
         }
-        unloadTopic(topicName, authoritative);
+
+        getPartitionedTopicMetadataAsync(topicName, authoritative, false)
+                .whenComplete((metadata, ex) -> {
+                    if (ex != null) {
+                        log.error("[{}] Failed to unload topic {}", clientAppId(), topicName);
+                        asyncResponse.resume(new RestException(ex));
+                    } else {
+                        if (metadata.partitions > 0) {
+                            final List<CompletableFuture<Void>> futures = Lists.newArrayList();
+
+                            for (int i = 0; i < metadata.partitions; i++) {
+                                TopicName topicNamePartition = topicName.getPartition(i);
+                                try {
+                                    futures.add(pulsar().getAdminClient().topics().unloadAsync(topicNamePartition.toString()));
+                                } catch (Exception e) {
+                                    log.error("[{}] Failed to unload topic {}", clientAppId(), topicNamePartition, e);
+                                    asyncResponse.resume(new RestException(e));
+                                    return;
+                                }
+                            }
+
+                            FutureUtil.waitForAll(futures).handle((result, exception) -> {
+                                if (exception != null) {
+                                    Throwable t = exception.getCause();
+                                    if (t instanceof NotFoundException) {
+                                        asyncResponse.resume(new RestException(Status.NOT_FOUND, t.getMessage()));
+                                    } else {
+                                        log.error("[{}] Failed to unload topic {}", clientAppId(), topicName, exception);
+                                        asyncResponse.resume(new RestException(exception));
+                                    }
+                                    return null;
+                                }
+
+                                asyncResponse.resume(Response.noContent().build());
+                                return null;
+                            });
+                        } else {
+                            validateAdminAccessForTenant(topicName.getTenant());
+                            validateTopicOwnership(topicName, authoritative);
+
+                            Topic topic = getTopicReference(topicName);
+                            try {
+                                topic.close(false).get();
```

Comment:
this should be implemented in an async way.

Reply:
OK